src/app/tabs/camera.cpp

```diff
@@ -0,0 +1,143 @@
+#include <app/config.hpp>
+#include <app/tabs/camera.hpp>
+#include <app/window.hpp>
+#include <app/theme.hpp>
+
+CameraTab::CameraTab(QWidget *parent) : QWidget(parent)
+{
+    QStackedLayout *stack = new QStackedLayout(this);
+    stack->addWidget(connect_widget());
+    stack->addWidget(cam_widget());
+
+    connect(this, &CameraTab::stream_connected, this, [stack]() {stack->setCurrentIndex(1);} );
+    connect(this, &CameraTab::stream_disconnected, this, [stack]() {stack->setCurrentIndex(0);} );
+
+    player = NULL;
+    // connect_stream(); TODO: offer auto-connect option
+}
+
+QWidget* CameraTab::cam_widget()
+{
+    QWidget* widget = new QWidget(this);
+    QVBoxLayout *layout = new QVBoxLayout(widget);
+
+    QHBoxLayout *connection = new QHBoxLayout;
+    status = new QLabel("Disconnected", widget);
+    status->setFont(Theme::font_16);
+    QPushButton *disconnect = new QPushButton("disconnect");
+    disconnect->setFont(Theme::font_14);
+    disconnect->setFlat(true);
+    disconnect->setIconSize(Theme::icon_36);
+    Theme::get_instance()->add_button_icon("wifi", disconnect);
+    connection->addWidget(status);
+    connection->addStretch();
+    connection->addWidget(disconnect);
+
+    connect(disconnect, &QPushButton::clicked, this, &CameraTab::disconnect_stream);
+
+    video_widget = new QVideoWidget;
+
+    layout->addWidget(video_widget);
+    layout->addLayout(connection);
+
+    return widget;
+}
+
+
+QWidget *CameraTab::connect_widget()
+{
+    QWidget *widget = new QWidget(this);
+    QVBoxLayout *layout = new QVBoxLayout(widget);
+
+    QLabel *label = new QLabel("connect camera stream for camera view", widget);
+    label->setFont(Theme::font_16);
+    label->setAlignment(Qt::AlignCenter);
+
+    Config *config = Config::get_instance();
+    QLineEdit *stream_url_input = new QLineEdit(config->get_cam_stream_url(), widget);
+    stream_url_input->setFixedWidth( 18*30 );

```

Comment:
remove spaces around parens (and add spaces between operator)

Reply:
hmm this might actually not work all the time

I've needed to avoid fixed sizes to not break the scaling (especially for smaller screens). You might want to throw this in a horizontal layout and throw some spacers on the sides

Reply:
i know cube has a screen thats like 480x234 or something like that :laughing:

his setup ends up upscaling that, but in terms of pixels yeah we might want to be conscious about that kind of stuff

Reply:
480x234???? I'd be impressed if it was useful at that resolution.

Even with spacers the box should have a minimum size. Other than doing a min() on the rendered font width and the screen I don't see how to ensure a reasonable number of characters being shown on normal screens. Seems like overkill for a normally-only-once-used URL field.

Reply:
well normally its not hah but he ends up upscaling it and it actually looks pretty decent

well you could always set the stretch of each component, so say your left and right spacers are 1/4 of the width each and the URL field is 1/2 (I've done that in other places)

I'm just not sure how the fixed width will be affected by the scaling, where i can see it possibly overflow out of the widget